File: go/mission/trust.go
package mission

// Capability trust boundary (Go port of trust.py). A capability whose Source is a third party is an
// execution surface; the Context Runtime treats untrusted sources like a plugin host: they stay
// DISCOVERABLE (the registry returns them) but are NOT bindable until the source is trusted — the
// bind door fails closed. Built-in capabilities (Source == "") are trusted, so nothing changes.

import (
	"bufio"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// TrustStore is the set of trusted capability sources, persisted one-per-line.
type TrustStore struct {
	Path    string
	trusted map[string]bool
}

func trustDefaultPath() string {
	if p := os.Getenv("AGENTIC_OS_TRUST_FILE"); p != "" {
		return p
	}
	home := os.Getenv("AGENTIC_OS_HOME")
	if home == "" {
		home, _ = os.UserHomeDir()
	}
	return filepath.Join(home, ".agentic-os", "trusted-sources")
}

// NewTrustStore loads the store from path (default AGENTIC_OS_TRUST_FILE/~/.agentic-os/trusted-sources).
func NewTrustStore(path string) *TrustStore {
	if path == "" {
		path = trustDefaultPath()
	}
	ts := &TrustStore{Path: path, trusted: map[string]bool{}}
	if f, err := os.Open(path); err == nil {
		defer f.Close()
		sc := bufio.NewScanner(f)
		for sc.Scan() {
			if s := strings.TrimSpace(sc.Text()); s != "" {
				ts.trusted[s] = true
			}
		}
	}
	return ts
}

// TrustStoreWith builds an in-memory store from an explicit set (no disk) — for tests/embedding.
func TrustStoreWith(sources ...string) *TrustStore {
	ts := &TrustStore{trusted: map[string]bool{}}
	for _, s := range sources {
		ts.trusted[s] = true
	}
	return ts
}

// IsTrusted: built-in/unsourced ("") is trusted; otherwise the source must be trusted.
func (ts *TrustStore) IsTrusted(source string) bool {
	return source == "" || ts.trusted[source]
}

func (ts *TrustStore) Trust(source string) {
	if source != "" {
		ts.trusted[source] = true
	}
}

func (ts *TrustStore) Revoke(source string) { delete(ts.trusted, source) }

func (ts *TrustStore) Save() error {
	if ts.Path == "" {
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(ts.Path), 0o755); err != nil {
		return err
	}
	keys := make([]string, 0, len(ts.trusted))
	for k := range ts.trusted {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return os.WriteFile(ts.Path, []byte(strings.Join(keys, "\n")+"\n"), 0o644)
}

File: agentic_os/mission/trust.py
"""Capability trust boundary — discover metadata-only, gate execution.

A capability whose ``source`` is a third party (a plugin, a cloned repo, an external registry) is an
execution surface: binding it lets it run. So the Context Runtime treats untrusted sources the way a
plugin host does (cf. xai-grok-build's plugin trust): capabilities from an untrusted source are still
**discovered and listed** (the registry returns them, metadata-only), but they are **not bindable**
until the source is trusted — the bind door fails closed. Built-in capabilities (``source == ""``)
are trusted by default, so nothing existing changes.

Trust is per-source (a canonical string), persisted one-per-line — the same shape as a plugin
trust-store — under ``AGENTIC_OS_TRUST_FILE`` or ``~/.agentic-os/trusted-sources``.
"""
from __future__ import annotations

import os
from pathlib import Path


def _default_path() -> Path:
    env = os.environ.get("AGENTIC_OS_TRUST_FILE")
    if env:
        return Path(env)
    home = os.environ.get("AGENTIC_OS_HOME") or os.path.expanduser("~")
    return Path(home) / ".agentic-os" / "trusted-sources"


class TrustStore:
    """The set of trusted capability sources. Built-in (empty source) is always trusted."""

    def __init__(self, path: str | os.PathLike | None = None, trusted: set[str] | None = None):
        self.path = Path(path) if path is not None else _default_path()
        self._trusted: set[str] = set(trusted) if trusted is not None else self._load()

    def _load(self) -> set[str]:
        try:
            return {ln.strip() for ln in self.path.read_text(encoding="utf-8").splitlines() if ln.strip()}
        except OSError:
            return set()

    def save(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.path.write_text("\n".join(sorted(self._trusted)) + "\n", encoding="utf-8")

    def is_trusted(self, source: str | None) -> bool:
        """Built-in/unsourced (``""``/None) is trusted; otherwise the source must be trusted."""
        return not source or source in self._trusted

    def trust(self, source: str, *, persist: bool = True) -> None:
        if source:
            self._trusted.add(source)
            if persist:
                self.save()

    def revoke(self, source: str, *, persist: bool = True) -> None:
        self._trusted.discard(source)
        if persist:
            self.save()

    def trusted(self) -> set[str]:
        return set(self._trusted)
